In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_style("whitegrid")

In [2]:
import pandas as pd

df = pd.read_csv(
    "../data/processed/fraud_processed.csv"
)

In [3]:
X = df.drop(
    columns=[
        'class',
        'signup_time',
        'purchase_time'
    ]
)

y = df['class']

In [4]:
X = X.drop(
    columns=[
        'user_id',
        'device_id'
    ],
    errors='ignore'
)

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    stratify=y,
    test_size=0.2,
    random_state=42
)

In [6]:
y_train.value_counts()

class
0    93502
1     9814
Name: count, dtype: int64

In [7]:
from imblearn.over_sampling import SMOTE

In [8]:
smote = SMOTE(
    random_state=42
)

X_train_res, y_train_res = smote.fit_resample(
    X_train,
    y_train
)

In [9]:
print(
    "Before SMOTE:"
)

print(
    y_train.value_counts()
)

print(
    "After SMOTE:"
)

print(
    y_train_res.value_counts()
)

Before SMOTE:
class
0    93502
1     9814
Name: count, dtype: int64
After SMOTE:
class
0    93502
1    93502
Name: count, dtype: int64


The training dataset exhibited a significant class imbalance, with 93,502 legitimate transactions and 9,814 fraudulent transactions. To address this issue, SMOTE (Synthetic Minority Oversampling Technique) was applied exclusively to the training set. After resampling, both classes contained 93,502 observations, creating a balanced dataset for model training while preserving the original distribution of the test set. SMOTE was chosen over undersampling because it retains all majority-class examples and reduces the risk of losing potentially valuable information.